### PR TITLE
Aptos and sui examples

### PR DESCRIPTION
**Pull request by moe-dev**
## Did you add a changeset?

If updating one of our packages, you'll likely need to add a changeset to your PR. To do so, run [`pnpm changeset`](https://pnpm.io/using-changesets#adding-new-changesets). `pnpm changeset` will generate a file where you should write a human friendly message about the changes. Note how this ([example](https://github.com/tkhq/sdk/blob/b409cd06790f011bf939adcf0755499b8e7497ae/.changeset/extra-http-exports.md?plain=1#L1)) includes the package name (should be auto added by the command) along with the type of [semver change (major.minor.patch)](https://semver.org/) (which you should set).

These changes will be used at release time to determine what packages to publish and how to bump their version. For more context see [this comment](https://github.com/tkhq/sdk/pull/67#issuecomment-1568838440).

### CODE REVIEW

File: examples/with-aptos/.env.local.example [hunk 1]
```diff
@@ -0,0 +1,6 @@
+API_PUBLIC_KEY="<Turnkey API Public Key (that starts with 02 or 03)>"
+API_PRIVATE_KEY="<Turnkey API Private Key>"
+BASE_URL="https://api.turnkey.com"
+ORGANIZATION_ID="<Turnkey organization ID>"
+APTOS_ADDRESS="<existing Aptos address in your organization>"
+APTOS_PUBLIC_KEY="<existing Aptos pubkey compressed matching your above address in your organization>"
```
Comment (andrewkmin) [inline examples/with-aptos/.env.local.example:6]: tiiiny nit: let's add an extra newline at the end

File: examples/with-sui/.env.local.example [hunk 2]
```diff
@@ -0,0 +1,6 @@
+API_PUBLIC_KEY="<Turnkey API Public Key (that starts with 02 or 03)>"
+API_PRIVATE_KEY="<Turnkey API Private Key>"
+BASE_URL="https://api.turnkey.com"
+ORGANIZATION_ID="<Turnkey organization ID>"
+SUI_ADDRESS="<existing Aptos address in your organization>"
+SUI_PUBLIC_KEY="<existing SUI pubkey compressed matching your above address in your organization>"
```
Comment (andrewkmin) [inline examples/with-sui/.env.local.example:6]: here too

File: examples/with-aptos/package.json [hunk 3]
```diff
@@ -0,0 +1,29 @@
+{
+  "name": "@turnkey/example-with-aptos",
+  "version": "0.1.0",
+  "private": true,
+  "scripts": {
+    "start": "pnpm tsx src/index.ts",
+    "advanced": "pnpm tsx src/advanced.ts",
```
Comment (andrewkmin) [inline examples/with-aptos/package.json:7]: let's remove `advanced`

File: examples/with-aptos/package.json [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    "clean": "rimraf ./dist ./.cache",
+    "typecheck": "tsc --noEmit"
+  },
+  "dependencies": {
+    "@inquirer/prompts": "^2.1.0",
+    "@noble/hashes": "1.4.0",
+    "@turnkey/api-key-stamper": "workspace:*",
```
Comment (andrewkmin) [inline examples/with-aptos/package.json:14]: do we need this directly?

File: examples/with-aptos/package.json [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    "@turnkey/http": "workspace:*",
+    "@turnkey/sdk-server": "workspace:*",
+    "aptos": "^1.21.0",
+    "async-retry": "^1.3.3",
+    "bs58": "^5.0.0",
```
Comment (andrewkmin) [inline examples/with-aptos/package.json:19]: or this? (and its accompanying `@types/bs58` dev dependency)

File: examples/with-aptos/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+  // Ensure r and s are 64 hex characters (32 bytes)
+  const rHex = r.padStart(64, "0");
+  const sHex = s.padStart(64, "0");
+
+  // Concatenate r and s to form the signature
+  const txSignatureHex = rHex + sHex;
+
+  // Validate signature length
+  if (txSignatureHex.length !== 128) {
+    console.error(
```
Comment (andrewkmin) [inline examples/with-aptos/src/index.ts:116]: we could just throw an error here

```
throw new Error("Invalid signature length for Ed25519. Expected 128 hex characters.") 
```

and the handler (`main().catch()...`) will catch the error and exit

File: examples/with-sui/package.json [hunk 7]
```diff
@@ -0,0 +1,29 @@
+{
+  "name": "@turnkey/example-with-sui",
+  "version": "0.1.0",
+  "private": true,
+  "scripts": {
+    "start": "pnpm tsx src/index.ts",
+    "advanced": "pnpm tsx src/advanced.ts",
```
Comment (andrewkmin) [inline examples/with-sui/package.json:7]: can remove

File: examples/with-sui/package.json [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    "clean": "rimraf ./dist ./.cache",
+    "typecheck": "tsc --noEmit"
+  },
+  "dependencies": {
+    "@inquirer/prompts": "^2.1.0",
+    "@mysten/sui.js": "^0.37.1",
+    "@noble/hashes": "1.4.0",
+    "@turnkey/api-key-stamper": "workspace:*",
```
Comment (andrewkmin) [inline examples/with-sui/package.json:15]: similar comments here with dependency cleanup

File: examples/with-sui/src/index.ts [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+
+  console.log(`Using Sui address: ${suiAddress}`);
+  const publicKeyBytes = Buffer.from(suiPublicKeyHex, "hex");
+
+  // Verify that the public key corresponds to the Sui address
+  const publicKey = new Ed25519PublicKey(publicKeyBytes);
+  const computedAddress = publicKey.toSuiAddress();
+
+  if (computedAddress !== suiAddress) {
+    console.error("The SUI_PUBLIC_KEY does not correspond to the SUI_ADDRESS.");
```
Comment (andrewkmin) [inline examples/with-sui/src/index.ts:49]: can also throw error here instead

File: examples/with-sui/src/index.ts [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+
+  if (coins.data.length === 0) {
+    console.error("No SUI coins found in the account.");
+    process.exit(1);
+  }
+
+  // Create and sign a transaction
+  const recipientAddress = await input({
+    message: "Recipient address:",
+    default: "<recipient_sui_address>",
```
Comment (andrewkmin) [inline examples/with-sui/src/index.ts:80]: we could create a SUI warchest that we default to here

File: examples/with-sui/src/index.ts [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+    `\nSending ${amount} MIST (${
+      Number(amount) / 1e9
+    } SUI) to ${recipientAddress}`
+  );
+
+  const tx = new TransactionBlock();
+
+  // Use the first coin for transfer
+
+  const reference_gas_price = await provider.getReferenceGasPrice();
```
Comment (andrewkmin) [inline examples/with-sui/src/index.ts:95]: tiny nit: would prefer camel case

**Review by andrewkmin — APPROVED** [review]
👏👏 overall lgtm, just some cleanup items

File: examples/with-sui/src/index.ts (hunk 10 above)
Comment (moe-dev) [reply examples/with-sui/src/index.ts:80]: It's mainnet so idk how I'd feel about that 😂

File: examples/with-aptos/src/index.ts [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+  );
+
+  // Prepare the transaction payload
+  const token = new TxnBuilderTypes.TypeTagStruct(
+    TxnBuilderTypes.StructTag.fromString("0x1::aptos_coin::AptosCoin")
+  );
+
+  const entryFunctionPayload =
+    new TxnBuilderTypes.TransactionPayloadEntryFunction(
+      TxnBuilderTypes.EntryFunction.natural(
```
Comment (andrewkmin) [inline examples/with-aptos/src/index.ts:62]: looking at this again, this is some interesting syntax lol. would be fun to explore the surface area of Aptos transactions at some point

File: examples/with-aptos/src/index.ts [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+    2000n, // Max gas amount
+    100n, // Gas price per unit
+    BigInt(Math.floor(Date.now() / 1000) + 600), // Expiration timestamp
+    new TxnBuilderTypes.ChainId(Number(chainId))
+  );
+
+  // Get the signing message
+  const signingMessage = TransactionBuilder.getSigningMessage(rawTxn);
+
+  // Sign the transaction using Turnkey with HASH_FUNCTION_NO_OP
```
Comment (andrewkmin) [inline examples/with-aptos/src/index.ts:95]:
```suggestion
  // Sign the transaction using Turnkey with HASH_FUNCTION_NOT_APPLICABLE (
```
we can include the context from here if we'd like: https://github.com/tkhq/sdk/blob/fa0f705cdc324023b8f78e146b10e3885ebebd47/packages/solana/src/index.ts#L151-L152

File: examples/with-aptos/src/index.ts [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+          BCS.bcsToBytes(
+            TxnBuilderTypes.AccountAddress.fromHex(recipientAddress)
+          ),
+          BCS.bcsSerializeUint64(Number(amount)),
+        ]
+      )
+    );
+
+  // Get account sequence number and chain ID
+  const [{ sequence_number: sequenceNumber }, chainId] = await Promise.all([
```
Comment (andrewkmin) [inline examples/with-aptos/src/index.ts:76]: don't need to do promises here but it'll work!

File: examples/with-sui/README.md [hunk 15]
```diff
@@ -0,0 +1,60 @@
+# Example: `with-sui`
+
+This is a simple example that walks through the following:
+
+- Construction of an sui transaction sending the funds out on mainnet
```
Comment (andrewkmin) [inline examples/with-sui/README.md:5]:
```suggestion
- Construction of a sui transaction sending the funds out on mainnet
```

File: examples/with-sui/src/index.ts (hunk 10 above)
Comment (andrewkmin) [reply examples/with-sui/src/index.ts:79]: LOLOL trueeee i take that back

**Review by andrewkmin — APPROVED** [review]
🏅

**Pull request merged**